### ISSUE TITLE
Implement a function for computing the number of messages received per MPI process

### ISSUE THREAD

**Issue by masterleinad** [dealii/dealii#6954]
As noted by @eldarkh  in https://github.com/dealii/dealii/pull/6851#issuecomment-401101542:
> On the other hand, and correct me if I am wrong here, I have noticed that when this function 
> [Utilities:MPI:compute_point_to_point_communication_pattern()] is used in deal.II it is only to find the
> number of processes to receive from, and this is what MPI_Reduce_scatter does in my 
> implementation, and this (without consequent sends/receives) is indeed noticeably faster than what 
> was originally implemented [...].

We should implement such a function.

**Comment by eldarkh** [comment]
I did an implementation of the new function that uses `MPI_Reduce` followed by `MPI_Scatter` to use with older versions of MPI and `MPI_Reduce_scatter_block` with MPI version 2.2 and above. Below is how timings look (I used 64 cores, and averaged the time over 10000 experiments, same communication pattern as in `point_to_point_pattern_01.cc`):

1. Time for Old implementation: 49.862 microsec
2. Time for `MPI_Reduce` followed by `MPI_Scatter`: 16.22 microsec
3. Time for `MPI_Reduce_scatter_block`: 10.243 microsec

So it is indeed beneficial to have this separate function and I can create a pull request as soon as I clean my code a bit and change all the function calls in the library.
